internal/grpc/example/server/main.go

```diff
@@ -0,0 +1,171 @@
+		// error code and message. Otherwise, any "anonymous" errors that don't implement the Status interface will be massaged
+		// into a Status object with code "Unknown" and handled appropriately by the gRPC library.
+		//
+		// See the following for more background and information:
+		// - https://avi.im/grpc-errors/#go
+		// - https://godoc.org/google.golang.org/grpc/codes
+		// - https://cloud.google.com/apis/design/errors (intended for Google developers, but generally applicable advice)
+		err := status.Errorf(codes.InvalidArgument, "We don't go to Ravenholm.")
+		return nil, err
+	}
+
+	// Now that we know that we have a "valid" location, we can safely return the response.
+	response := &pb.WeatherResponse{
+		Description: "It's always Sunny!",
+		Temperature: 25.0,
+	}
+
+	// Note: The gRPC library handles serializing the response to the client - we can return the response struct directly.
+	return response, nil
+}
+
+// SubscribeWeatherAlerts is a Server Streaming (aingle request, multiple responses) RPC that returns a stream of relevant weather alerts.
+func (s *weatherServer) SubscribeWeatherAlerts(req *pb.AlertRequest, stream pb.WeatherService_SubscribeWeatherAlertsServer) error {
+	ctx := stream.Context()
+	for i := 0; i < 5; i++ {
+		select {
+		case <-ctx.Done(): // The client either explicitly canceled the operation or the context timed out.
+
+			// status.FromContextError is a convenience function that converts a context error
+			// to a Status object (with code codes.Canceled or codes.DeadlineExceeded).
+			//
+			// Returning a proper status (instead of a nil error) here makes it clearer to the caller / service-wide observability tools that look at
+			// response codes what exactly happened with this RPC call.
+			return status.FromContextError(ctx.Err()).Err()
+
+		case <-time.After(2 * time.Second): // Wait 2 seconds between the fake alerts we're sending.
+
+			// Send a message to the client
+			err := stream.Send(&pb.AlertResponse{
+				Alert: "Severe weather alert!",
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// UploadWeatherData is a long-running Client Streaming RPC (multiple request messages) is that is used to receive weather sensor data from a client.
+func (s *weatherServer) UploadWeatherData(stream pb.WeatherService_UploadWeatherDataServer) error {
+	ctx := stream.Context()
+	for {
+		if err := ctx.Err(); err != nil {
+			// The client either explicitly canceled the operation, or the deadline expired.
+			return status.FromContextError(err).Err()
+		}
+
+		data, err := stream.Recv()
+		if errors.Is(err, io.EOF) {
+			// io.EOF is a sentinel value that indicates that the client has explicitly closed its end of the stream, which signals the end of the RPC.
+
+			// We can use SendAndClose to send a final message to the client and close our end of the stream.
+			return stream.SendAndClose(&pb.UploadStatus{
+				Message: "Data received successfully",
+			})
```

Comment:
Is it legitimate to just call Close as well? In case there's not really a message other than "no error" we want to send.
If that's true, maybe we could add 
```suggestion
			return stream.SendAndClose(&pb.UploadStatus{
				Message: "Data received successfully",
			})
			// Or: return stream.Close() to simply close the connection without triggering an error on the client side.
```